internal/pkg/remote/client.go

```diff
@@ -195,24 +204,23 @@ func (c *Client) Send(
 			}
 		}
 
-		requester.lastUsed = time.Now().UTC()
-
 		resp, err = requester.client.Do(req.WithContext(ctx))
-		if err != nil {
-			requester.lastErr = err
-			requester.lastErrOcc = time.Now().UTC()
 
+		// Using the same lock that was used for sorting above
+		c.clientLock.Lock()
+		requester.SetLastError(err)
```

Comment:
I'm not convinced this is completely correct if `Send()` is called concurrently for two different API endpoints on the same host. 

This assumes that errors from one HTTP route have equal weight for other routes on the same host which isn't actually true.  It might be true if the error is a 5xx error but isn't necessarily true for 4xx.

At the same time I'm not sure that fixing this makes any functional difference since this just alters the order in which hosts are tried. 

This is probably worth a comment though in case the logic here changes in the future. Can you add some comments making it obvious that this function is intended to be used concurrently for multiple APIs on the same host?

Reply:
The only obvious fixes for this are having the errors be per API, having the agent use two clients, etc. Those seems like they add complexity vs just accepting that under some circumstances the priority of the hosts won't be exactly optimal when we are sorting them.

Reply:
Not debating the existing code logic here. The logic of the errors "per client" is still the same, and I'm not exactly sure why this was implemented the way it is, just making the existing code safe for using concurrently.